File: java/JaeHyeon/pro_86971.java
import java.util.*;

public class pro_86971 {

    static List<List<Integer>> tree = new ArrayList<>();
    static int answer = Integer.MAX_VALUE;

    static public int solution(int n, int[][] wires) {

        for (int i = 0; i <= n; i++) {
            tree.add(new ArrayList<>());
        }

        for (int[] wire : wires) {
            tree.get(wire[0]).add(wire[1]);
            tree.get(wire[1]).add(wire[0]);
        }

        for (int i = 0; i < wires.length; i++) {

            boolean[] visit = new boolean[n+1];

            tree.get(wires[i][0]).remove(Integer.valueOf(wires[i][1]));
            tree.get(wires[i][1]).remove(Integer.valueOf(wires[i][0]));

            int count = dfs(1, visit);

            answer = Math.min(answer, Math.abs(count - (n - count)));

            tree.get(wires[i][0]).add(wires[i][1]);
            tree.get(wires[i][1]).add(wires[i][0]);
        }

        return answer;
    }

    private static int dfs(int start, boolean[] visit) {
        visit[start] = true;
        int cnt = 1;

        for (int nextVertex : tree.get(start)) {
            if (!visit[nextVertex]) {
                cnt += dfs(nextVertex, visit);
            }
        }

        return cnt;
    }

    public static void main(String[] args) {
        System.out.println(solution(9, new int[][]{{1,3},{2,3},{3,4},{4,5},{4,6},{4,7},{7,8},{7,9}}));
    }
}

File: soya/src/pro_86971.ts
function solution(n, wires) {
  var answer = n;

  //1번 기준으로 갈곳 없을때까지 탐색
  //탐색 종료 후 두 전력망 송전탑 개수 차이 = n - cnt
  function dfs(curT, cuttedWires, cnt, visited) {
    for (let wire of cuttedWires) {
      if (wire[0] === curT && !visited[wire[1]]) {
        visited[wire[1]] = true;
        dfs(wire[1], cuttedWires, cnt + 1, visited);
      }
      if (wire[1] === curT && !visited[wire[0]]) {
        visited[wire[0]] = true;
        dfs(wire[0], cuttedWires, cnt + 1, visited);
      }
    }
  }

  //하나씩 선을 잘라서 탐색
  for (let i = 0; i < wires.length; i++) {
    let cuttedWires = [...wires];
    let visited = new Array(wires.length + 1).fill(false); //1~n
    cuttedWires.splice(i, 1);
    //1번 송전탑 기준으로 연결된 송전탑 탐색
    visited[1] = true;
    dfs(1, cuttedWires, 1, visited);
    //방문체크된 애들 카운트해서 정답 적기
    let count = 0;
    for (let visit of visited) {
      if (visit) count++;
    }
    answer =
      Math.abs(n - 2 * count) < answer ? Math.abs(n - 2 * count) : answer;
  }

  return answer;
}
